cpp/src/sender.cpp

```diff
@@ -60,14 +61,20 @@ class target sender::target() const {
 }
 
 namespace {
-// TODO: revisit if thread safety required
-uint64_t tag_counter = 0;
+std::atomic<uint64_t> tag_counter(0);
```

Comment:
This would of course work. I was also thinking about maybe making the counter be an instance variable of a `sender`. Let's see what others think first.

Reply:
I would probably lean toward a sender-specific value as well, though both ways have their own advantages.

@astitcher ? @ssorj ?

Reply:
It would make a lot of sense (from a performance stand point) to have the tag counter be scoped to the connection in some way (as this is the serialisation granule of proton).
Is it true that the delivery tags are only scoped to the sender then. So that different senders can use identical tags? if so we should just make this connection or more granularly scoped and just zero out the count when we create a new connection/session/sender/etc.

Reply:
To be clear if we can make the tag counter scoped to the connection or lower then we don't need any thread synchronisation at all. This is because it is an underlying assumption in proton that a connection can only be processed on a single thread at a time.

Reply:
> Is it true that the delivery tags are only scoped to the sender then. So that different senders can use identical tags?

Yes, tags are only used in the scope of a particular link, so they only need to be unique for the outstanding set of unsettled deliveries on that link:

"The delivery-tag MUST be unique amongst all deliveries that could be considered unsettled by either end of the link. "
http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#doc-idp438000

Reply:
>To be clear if we can make the tag counter scoped to the connection or lower then we don't need any thread synchronisation at all. This is because it is an underlying assumption in proton that a connection can only be processed on a single thread at a time.

Yep, thats what pompted my original reply on the other PR, when I thought the variable was sender-specific, before it was pointed out it was actually a global variable.

Reply:
I've made the changes. Can you please check if this is what you meant?

Reply:
It's what I meant. Let's wait what Andrew thinks, as the cpp binding maintainer...